Introduce build environment

davityavryan opened the pull request:
- I have added `rollup` and `babel`.
- Build is done via `yarn` (I hope you don't mind)

- [ ] Browserslist should be re-defined
- [x] Fix Browserslist env targets. at the moment there are `production` and `development` but we use different `BABEL_ENV`-s.
- [x] Hierarchy needs to be preserved ~Understand why there is a need for having `core` folder. What's the plan? if there is no need, then drop it~
- [x] 2 spaces ~Spaces, can I define 4 instead of 2 if you don't mind?~

After this you just need to run `yarn version --patch`, `yarn build` and `yarn publish ./build`

davityavryan committed:
Refactor README location and add LICENSE

davityavryan committed:
Add build environment

davityavryan commented on .browserslistrc:
```diff
@@ -0,0 +1,11 @@
+[production test]
```
needs to be reviewed

davityavryan commented:
@andregardi I can also add #34 functionality. Of course it will required to do some minor fixes.

andregardi commented:
@davityavryan , Thank you for the great work you are doing here!

1 - Thank you for introducing me to rollup, great tool!

2 - Do we really need `yarn `here? Could we use something like: 
`"build": "npm-run-all build:umd build:cjs build:esm build:es build:copy-files"`

3 - About the `core` folder:
I am working on some optional plugins to make this lib more useful. The core folder will help me to implement a tree-shake system, allowing the app to only import what the project needs. There are some cool features, like integration with immerjs that will allow us to mutate the state and still get the components to be updated. 

4 - About lint and 4 spaces:
I like the idea of having a lint on the project. But maybe it is too much for the same pull request.
I really prefer 2 spaces over 4 and think that the airbnb rules are too restritive, but I am open to talk about it.

5 - Can we add an example on how to import this other builds, like the `umd`, from our package?
Will it be something like 'use-global-hook/umd' ?
Is there also a guide on when should use it, like regular import to target chrome and `umd `for edge?
Regular, `cjs`, `ems`, `es`, `umd`. Do we need all that versions? It is ok if we need, but it is no very clear to me.

6 - Why do we need `@babel/runtime` on `peerDependencies`?
The app is going to need it installed even if we are not using the `ES5 `version?

davityavryan commented:
@andregardi  you are welcome. I was using and recommending your package several months. So thanks to you for creating it.

1. Sure, but as you can see I am using it only for UMD at the moment. reason for that is that when I am producing CommonJS and ES versions with them they are merged into one file. Don't think that it is a big deal but I just wanted to preserve hierarchy  under `src` folder. I will try to fix that and either use rollup for all variants or just babel transpilation.

2. No it's not mandatory. and it is not for chaining build scripts. as you can see it produced lock file and `yarn` is better than `npm` in a various ways (see [YARN vs. NPM](https://medium.com/@j.dumadag718/yarn-vs-npm-b2d58289fb9b). So although it's your package and you are deciding which one to use. I highly recommend to use yarn.

3. Got it. then I will need to preserve that structure inside `build` folder.

4. No problem, 2 spaces it is.

5. Ah yeah see [What are CJS, AMD, UMD, and ESM in Javascript?](https://dev.to/iggredible/what-the-heck-are-cjs-amd-umd-and-esm-ikm) for more info. It's not that for Edge you need to target UMD or other. tools like Webpack will do it's job depending on your targets. Usually they take CJS version. but if your browsers are modern they try to use ESM. UMD is better for embedding in html right away (like Codepen example). but If you don't plan to have this, we can just remove.

6. At the moment ESM code contains imports to that and if we will not mention it in `peerDependencies` then it end user's build might fail(although most users use Babel transilation in which case they will have runtime installed, but yeah).  But I will try to inline it.
![image](https://user-images.githubusercontent.com/8176311/89734173-cb3e7000-da5a-11ea-8813-9e5701bef9b0.png)

davityavryan commented:
p.s. Worth reading [Code-splitting for libraries—bundling for npm with Rollup 1.0](https://levelup.gitconnected.com/code-splitting-for-libraries-bundling-for-npm-with-rollup-1-0-2522c7437697)

davityavryan commented on package.json:
```diff
@@ -1,12 +1,43 @@
 {
-  "license": "MIT",
-  "main": "src/core/index.js",
+  "version": "0.1.13",
```
should be bumped to `1.0.0` as we have breaking changes

davityavryan commented:
@andregardi now everything is under rollup and I think you will like the hierarchy. ;)

andregardi commented:
@davityavryan 
I made some tests with the new build and it seems to work well on the examples that we already have.
I think we are going to release a new version on the weekend with that.

By the way, I am not sure if we have breaking changes since the examples are still running without a code change. Can you explain to me?

davityavryan commented:
@andregardi we have changed the places of files and most importantly we have introduced CJS and ESM versions, so Build tools like Webpack will use different version after update.
For me that sound like Breaking changes. Not minor, not patch but major version update.

So I would ask you to run `yarn version --major`, `yarn build` and `yarn publish ./build`. And please don't forget to do `git push --tags` ;)

see [Semver](https://semver.org/) definition for each one.

davityavryan commented:
p.s. we can add tests to repo so that they will be automated. if you want I can help with that too. ;)
